include/llama/Copy.hpp

```diff
@@ -0,0 +1,364 @@
+        constexpr auto LanesDst = internal::aosoaLanes<DstMapping>;
+
+        if (srcView.mapping.arrayDims() != dstView.mapping.arrayDims())
+            throw std::runtime_error{"Array dimensions sizes are different"};
+
+        constexpr auto srcIsAoSoA = LanesSrc != std::numeric_limits<std::size_t>::max();
+        constexpr auto dstIsAoSoA = LanesDst != std::numeric_limits<std::size_t>::max();
+
+        static_assert(srcIsAoSoA || dstIsAoSoA, "At least one of the mappings must be an AoSoA mapping");
+        static_assert(!srcIsAoSoA || decltype(srcView.storageBlobs)::rank == 1);
+        static_assert(!dstIsAoSoA || decltype(dstView.storageBlobs)::rank == 1);
+
+        const auto arrayDims = dstView.mapping.arrayDims();
+        const auto flatSize
+            = std::reduce(std::begin(arrayDims), std::end(arrayDims), std::size_t{1}, std::multiplies<>{});
+
+        // the same as AoSoA::blobNrAndOffset but takes a flat array index
+        auto mapAoSoA = [](std::size_t flatArrayIndex, auto coord, std::size_t Lanes) LLAMA_LAMBDA_INLINE
+        {
+            const auto blockIndex = flatArrayIndex / Lanes;
+            const auto laneIndex = flatArrayIndex % Lanes;
+            const auto offset = (sizeOf<RecordDim> * Lanes) * blockIndex + offsetOf<RecordDim, decltype(coord)> * Lanes
+                + sizeof(GetType<RecordDim, decltype(coord)>) * laneIndex;
+            return offset;
+        };
+        // the same as SoA::blobNrAndOffset but takes a flat array index
+        auto mapSoA = [&](std::size_t flatArrayIndex, auto coord, bool mb) LLAMA_LAMBDA_INLINE
+        {
+            const auto blob = mb * flatRecordCoord<RecordDim, decltype(coord)>;
+            const auto offset = !mb * offsetOf<RecordDim, decltype(coord)> * flatSize
+                + sizeof(GetType<RecordDim, decltype(coord)>) * flatArrayIndex;
+            return NrAndOffset{blob, offset};
+        };
+
+        auto mapSrc = [&srcView, &mapAoSoA, &mapSoA](std::size_t flatArrayIndex, auto coord) LLAMA_LAMBDA_INLINE
+        {
+            if constexpr (srcIsAoSoA)
+                return &srcView.storageBlobs[0][0] + mapAoSoA(flatArrayIndex, coord, LanesSrc);
+            else
+            {
+                const auto [blob, off] = mapSoA(flatArrayIndex, coord, MBSrc);
+                return &srcView.storageBlobs[blob][off];
+            }
+        };
+        auto mapDst = [&dstView, &mapAoSoA, &mapSoA](std::size_t flatArrayIndex, auto coord) LLAMA_LAMBDA_INLINE
+        {
+            if constexpr (dstIsAoSoA)
+                return &dstView.storageBlobs[0][0] + mapAoSoA(flatArrayIndex, coord, LanesDst);
+            else
+            {
+                const auto [blob, off] = mapSoA(flatArrayIndex, coord, MBDst);
+                return &dstView.storageBlobs[blob][off];
+            }
+        };
+
+        constexpr auto L = std::min(LanesSrc, LanesDst);
+        static_assert(!srcIsAoSoA || LanesSrc % L == 0);
```

Comment:
Can this fail for user code? If so, this needs an error message.

Reply:
Yes it can and you are right!

Reply:
I replaced the code by a different implementation which does not have this problem.